### ISSUE TITLE
Support for code regions

### ISSUE THREAD

**Issue by Sam-bit** [dart-lang/language#285]
Can we have code regions facility for dart just like #region in c# for structural code blocks purpose. And thus can be used in flutter also

**Comment by munificent** [comment]
My feeling has long been that `#region` is a somewhat useful feature in C#, but also implies there are other problems in the code. If a class is so big it needs to be broken down into "chapters", it's probably just too big.

For C#, the argument was that much of that code was generated WinForms code, so it made sense to be able to hide it. But I think partial classes (which if I recall came later) were a cleaner solution because it let you move that stuff to a separate physical file. That makes it much easier to integrate with code generators and otherwise manage things more cleanly.

For Dart, in many cases you can use mixins to break up a giant class into smaller more manageable pieces. You can even, if you want, put those all into the same file, which then also plays nice with library privacy.

I think that would be a better approach than applying a band-aid like `#region`.

**Comment by listepo** [comment]
I think this is solved with a plugin for the IDE

**Comment by ercankayaNet** [comment]
which plugin?

**Comment by arbaieffendi** [comment]
This features would be useful

**Comment by techfan101** [comment]
Here's another vote for code regions as a useful feature.

While I agree with @munificent regarding classes that grow too large, I don't know that code regions actually encourage that growth.  Sometimes bad developers are bad no matter how hard you wish they were better :)

Also, in some cases you may not have total control over the organization of your class.  For example, you may be implementing a particularly large interface from an external package.  In these cases, code regions can at least help organize the clutter from that (sometimes poorly designed) interface.

**Comment by anasnaguib** [comment]
This Feature will be very useful in managing parts for class, or parts of code written by different team members.

**Comment by mika76** [comment]
I would also love this, although I think it really only needs to be part of the vscode extension, not specifically the sdk?

> which plugin?
https://marketplace.visualstudio.com/items?itemName=maptz.regionfolder

**Comment by munificent** [comment]
Yeah, it's definitely possible for an IDE to support regions just by recognizing comments that contain some special marker text. In that case, no language support is needed.
